### PR TITLE
Add proxy support to docker, cri_containerd and kubelet

### PR DESCRIPTION
**Pull request by nickperry**
This PR adds the ability to configure proxying for the container runtime and Kubelet.
    
    Add 3 new string variables - http_proxy, https_proxy and no_proxy.

    [*http_proxy*]
     Configure the HTTP_PROXY environment variable

    [*https_proxy*]
     Configure the HTTPS_PROXY environment variable

    [*no_proxy*]
     Configure the NO_PROXY environment variable

    Add a boolean 'container_runtime_use_proxy' which when set to true will cause
    the new proxy variables to be applied to the container runtime.

    [*container_runtime_use_proxy*]
     Configure whether the container runtime should be configured to use a proxy.
     If set to true, the container runtime will use the http_proxy, https_proxy and
     no_proxy values.

    Add a boolean 'kubelet_use_proxy' which when set to true will cause the
    http_proxy, https_proxy and no_proxy variables to be applied to the kubelet.

    [*kubelet_use_proxy*]
     Configure whether the kubelet should use a proxy. If set to true, the kubelet
     will use the http_proxy, https_proxy and no_proxy values

### CODE REVIEW

**Comment by nickperry** [comment]
@deric I cherry-picked your Don't try to guess docker_version commit, but still the spec tests are failing for me. The code I've added works great though in my envs, for containerd, docker and kubelet. Any idea what is going wrong with the spec tests?

File: manifests/init.pp [hunk 1]
```diff
@@ -661,16 +682,21 @@
...
-  String $docker_log_max_size                             = '100m',
-  Boolean $disable_swap                                   = true,
-  Boolean $manage_kernel_modules                          = true,
-  Boolean $manage_sysctl_settings                         = true,
-  Boolean $create_repos                                   = true,
-  String $image_repository                                = 'k8s.gcr.io',
-  Array[String] $default_path                             = ['/usr/bin', '/usr/sbin', '/bin', '/sbin', '/usr/local/bin'],
-  String $cgroup_driver                                   = $facts['os']['family'] ? {
+  Optional[String] $docker_extra_daemon_config                   = undef,
+  Optional[String] $http_proxy                                   = $undef,
```
Comment (deric) [inline manifests/init.pp:686]: @nickperry here's extra `$`, `undef` is a type, `$undef` would be a variable (but undefined :slightly_smiling_face: )

Comment (nickperry) [reply manifests/init.pp:686]: 🤦‍♂️Thank you so much! 🙂 I completely failed to see that.

**Comment by nickperry** [comment]
@pmcmaw @chelnak 

Sorry to ask directly, but do you think it'll be much longer before this is reviewed? We've been running off our fork for a while as we need this feature, but I'd like to tidy up and move our environments back to the upstream module.

**Review by pmcmaw — APPROVED** [review]
Hey @nickperry, no worries. We aim to have a look at PRs on Mondays during community day but this week was super busy so no one in the team got around to having a look. I have had a look now and looks like a great PR. Always great to see tests and docs without highlighting. I'm happy to merge, but please be aware that this module has quite a few outstanding PRs etc, therefore I cannot put a timeline on when this will officially be released to the Forge, but will be available in `main`.

**Pull request merged**